pkg/storage/stores/shipper/compactor/deletion/delete_request.go

```diff
@@ -0,0 +1,90 @@
+package deletion
+
+import (
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/loki/pkg/storage/stores/shipper/compactor/retention"
+)
+
+// DeleteRequest holds all the details about a delete request.
+type DeleteRequest struct {
+	RequestID string              `json:"request_id"`
+	UserID    string              `json:"-"`
+	StartTime model.Time          `json:"start_time"`
+	EndTime   model.Time          `json:"end_time"`
+	Selectors []string            `json:"selectors"`
+	Status    DeleteRequestStatus `json:"status"`
+	Matchers  [][]*labels.Matcher `json:"-"`
+	CreatedAt model.Time          `json:"created_at"`
+}
+
+func (d *DeleteRequest) IsDeleted(entry retention.ChunkEntry) (bool, []model.Interval) {
+	if d.UserID != unsafeGetString(entry.UserID) {
+		return false, nil
+	}
+
+	if !intervalsOverlap(model.Interval{
+		Start: entry.From,
+		End:   entry.Through,
+	}, model.Interval{
+		Start: d.StartTime,
+		End:   d.EndTime,
+	}) {
+		return false, nil
+	}
+
+	matchers := make([][]*labels.Matcher, len(d.Selectors))
+	var err error
+
+	for j, selector := range d.Selectors {
+		matchers[j], err = parser.ParseMetricSelector(selector)
+
+		if err != nil {
+			return false, nil
+		}
+	}
+
+	matches := false
+	for _, matchers := range matchers {
+		if labels.Selector(matchers).Matches(entry.Labels) {
+			matches = true
+			break
+		}
+	}
+
+	if !matches {
+		return false, nil
+	}
+
+	if d.StartTime <= entry.From && d.EndTime >= entry.Through {
```

Comment:
Can you start by checking if the entry time match the request first and exit quickly if not.

This will be much faster.

Reply:
Ha nevermind this is what you do.